drivers/sensor/ds18b20/ds18b20.c

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright (c) 2018 Roman Tataurov <diytronic@yandex.ru>
+ * Copyright (c) 2022 Thomas Stranger
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * Driver for DS18B20 1-Wire temperature sensors
+ * A datasheet is available at:
+ * https://datasheets.maximintegrated.com/en/ds/DS18B20.pdf
+ */
+#define DT_DRV_COMPAT maxim_ds18b20
+
+#include <drivers/sensor.h>
+#include <drivers/sensor/w1_sensor.h>
+#include <kernel.h>
+#include <logging/log.h>
+#include <sys/__assert.h>
+
+#include "ds18b20.h"
+
+LOG_MODULE_REGISTER(DS18B20, CONFIG_SENSOR_LOG_LEVEL);
+
+static int ds18b20_configure(const struct device *dev);
+
+/* measure wait time for 9-bit, 10-bit, 11-bit, 12-bit resolution respectively */
+static const uint16_t measure_wait_ms[4] = { 94, 188, 376, 750 };
+
+static inline void ds18b20_temperature_from_raw(uint8_t *temp_raw,
+						struct sensor_value *val)
+{
+	uint16_t temp = sys_get_le16(temp_raw);
+
+	val->val1 = (temp & 0x0FFF) >> 4;
+	val->val2 = (temp & 0x000F) * 10000;
```

Comment:
I think this is wrong.  My variant of calculating the fractional part of the temperature: 
`val->val2 = ((temp & 0x000F) * 1000000) >> 4;`

Reply:
@KondratevAndrei Thanks for reporting this. We came across the same issue while using it.

Reply:
Since the value in **scratchpad.temp** is stored as a signed 16 bit integer and **sys_get_le16()** returns unsigned value, we have to do this:
`int16_t temp = (int16_t)sys_get_le16(temp_raw);`
in other way, the tempere values below 0 were wrong for me.
This works well for me both with positive and negative values:
```
	int16_t temp = (int16_t)sys_get_le16(temp_raw);
	val->val1 = temp / 16;
	val->val2 = temp % 16 * 1000000 >> 4; 
```

Reply:
The existing temperature sensors in drivers/sensor (the tmpxxx family for instance) simply multiplies by a scaling factor and then do a divide by 1000000 and modulus 1000000. That will work fine here as well, there is no need for the rightshift.

```
    int16_t temp = (int16_t)sys_get_le16(temp_raw);
    int32_t microdeg = (int32_t)temp * DS18B20_TEMP_SCALE;

    val->val1 = microdeg / 1000000;
    val->val2 = microdeg % 1000000;
```

where `DS18B20_TEMP_SCALE` should be defined in ds18b20.h:
```
#define DS18B20_TEMP_SCALE (1000000/16)
```